src/Core/Grid/Presenter/GridPresenter.php

```diff
@@ -0,0 +1,116 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Presenter;
+
+use PrestaShop\PrestaShop\Core\Grid\Column\ColumnCollectionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Column\ColumnInterface;
+use PrestaShop\PrestaShop\Core\Grid\GridInterface;
+
+/**
+ * Class GridPresenter is responsible for presenting grid
+ */
+final class GridPresenter implements GridPresenterInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function present(GridInterface $grid)
+    {
+        $definition = $grid->getDefinition();
+        $searchCriteria = $grid->getSearchCriteria();
+        $data = $grid->getData();
+
+        return [
+            'id' => $definition->getId(),
+            'name' => $definition->getName(),
+            'filter_form' => $grid->getFilterForm()->createView(),
+            'columns' => $definition->getColumns()->toArray(),
+            'actions' => [
+                'panel' => $definition->getPanelActions()->toArray(),
+                'bulk' => $definition->getBulkActions()->toArray(),
+            ],
+            'data' => [
+                'rows' => $this->presentRows($grid),
+                'rows_total' => $data->getRowsTotal(),
+                'query' => $data->getQuery(),
+            ],
+            'pagination' => [
+                'offset' => $searchCriteria->getOffset(),
+                'limit' => $searchCriteria->getLimit(),
+            ],
+            'sorting' => [
+                'order_by' => $searchCriteria->getOrderBy(),
+                'order_way' => $searchCriteria->getOrderWay(),
+            ],
+        ];
+    }
+
+    /**
+     * Present grid data
+     *
+     * @param GridInterface $grid
+     *
+     * @return array
+     */
+    private function presentRows(GridInterface $grid)
+    {
+        $presentedRows = [];
+
+        $rows = $grid->getData()->getRows();
+        $columns = $grid->getDefinition()->getColumns();
+
+        foreach ($rows as $row) {
+            $rowData = $this->applyColumnModifications($row, $columns);
+
+            $presentedRows[] = $rowData;
+        }
+
+        return $presentedRows;
+    }
+
+    /**
+     * Some columns may modify row data
+     *
+     * @param array                     $row
+     * @param ColumnCollectionInterface $columns
+     *
+     * @return array
+     */
+    private function applyColumnModifications(array $row, ColumnCollectionInterface $columns)
+    {
+        /** @var ColumnInterface $column */
+        foreach ($columns as $column) {
+            if (!is_callable($column->getModifier())) {
```

Comment:
what is the need of a modifier here? I'm not sure to understand what is the need?

Shouldn't be a hook?

Reply:
initial idea of `modifier` was to be able to modify raw data. Example: you have price column in your grid and this price is raw value from db, like 25.59988, so modifiers job would be to format this price for display, so it could become '25.60 EUR'. We still use most simple template to render it, but the value is changed by modifier. 

@mickaelandrieu  do you understand it now? if so, maybe its not needed or you have other idea? :)